### PR TITLE
WIP KBase information for splitting fluentd

### PR DESCRIPTION
**Pull request by ewolinetz**
@jcantrill @sosiouxme @timothysc @richm @t0ffel PTAL

I need to still test the whole solution e2e, but I'm looking for fact checking/word-smithing as the intent for this is to also live in a KBase article.

### CODE REVIEW

File: deployer/README.md [hunk 1]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
     logging     kibana-ops   kibana-ops.example.com              logging-kibana-ops
 
 (In this example there are no overlapping routes.)
+
+## Splitting your Fluentd for high scale capacity (Advanced)
```
Comment (portante) [inline deployer/README.md]: Nice, though this seems like a lot of steps, and providing an ansible playbook which does all that is described here, just adding a set of variables, would be really powerful.

File: deployer/README.md [hunk 2]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+    $ oc label node --all logging-infra-fluentd-
+
+Where "logging-infra-fluentd" is the nodeSelector configured for the Fluentd daemonset.
+
+Next, create the configmap that will be used for the Enriching Fluentd.
+
+First, generate a random key that will be used for communication between the two
+types of Fluentd pods:
+
+    $ tr -dc 'a-zA-Z0-9' < /dev/urandom | head -c 128
```
Comment (portante) [inline deployer/README.md]: You might want to give this a shell variable name for the value, as a way to refer to it in the .conf file below.

File: deployer/README.md [hunk 3]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+    ## source
+    <source>
+      @type secure_forward
+      @label @INGRESS
+
+      self_hostname ${HOSTNAME}-enrichment
+      bind 0.0.0.0
+      port 24284
+
+      shared_key <our generated string>
```
Comment (portante) [inline deployer/README.md]: Would it be worthwhile to say, "our generated secret key from above"?  And potentially reference the shell variable name here?

File: deployer/README.md [hunk 4]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+      @include configs.d/dynamic/input-syslog-*.conf
+      @include configs.d/openshift/input-post-*.conf
+    ##
+
+    <label @INGRESS>
+      <match **>
+        @type secure_forward
+
+        self_hostname ${HOSTNAME}
+        shared_key <our generated string>
```
Comment (portante) [inline deployer/README.md]: Would it be worthwhile to say, "our generated secret key from above"? And potentially reference the shell variable name here?

File: deployer/README.md [hunk 5]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+      </match>
+    </label>
+
+Next, since our collecting Fluentd expects to ship its logs to a service called
+'logging-enrichment-fluentd' that service will need to be created:
+
+    $ oc expose dc logging-enrichment-fluentd --port=24284
+
+At this point, the enrichment Fluentd pods can be scaled up. They do not need to
+be 1:1 with the Fluentd collectors, A ratio of 1:10 is likely a good start:
```
Comment (portante) [inline deployer/README.md]: Note that the fluentd process is not multi-threaded, so you might want to consider using http://docs.fluentd.org/articles/in_multiprocess in order to make use of the cores available to the pod.

File: deployer/README.md [hunk 6]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
     logging     kibana-ops   kibana-ops.example.com              logging-kibana-ops
 
 (In this example there are no overlapping routes.)
+
+## Splitting your Fluentd for high scale capacity (Advanced)
+
+In scenarios where the number of Fluentd nodes starts to exceed 100, the watching
```
Comment (t0ffel) [inline deployer/README.md]: 1. 'Fluentd nodes' - sounds weird to me. 'nodes running Fluentd' i think is better.
2. From what I understand each fluentd starts 2 watches. What's the default value for maxRequestsInFlight? 400?

File: deployer/README.md [hunk 7]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+          name: enrichment-fluentd-elasticsearch
+          labels:
+            provider: openshift
+            component: "fluentd"
+        spec:
+          terminationGracePeriodSeconds: 300
+          serviceAccountName: aggregated-logging-fluentd
+          containers:
+          - name: enrichment-fluentd-elasticsearch
+            image: ${IMAGE_PREFIX}logging-fluentd:${IMAGE_VERSION}
```
Comment (t0ffel) [inline deployer/README.md]: this seem to have been taken from the template and needs to be updated.

I however think that it is better to provide template for the DC as well as for the service

File: deployer/README.md [hunk 8]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+            imagePullPolicy: Always
+            securityContext:
+              privileged: true
```
Comment (t0ffel) [inline deployer/README.md]: this container shouldn't be privileged

File: deployer/README.md [hunk 9]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+              limits:
+                cpu: 100m
+            ports:
+              - containerPort: 1095
+                hostPort: 1095
+                protocol: TCP
+                purpose: "to prevent more than one per node"
+                volumeMounts:
+                - name: runlogjournal
+                  mountPath: /run/log/journal
```
Comment (t0ffel) [inline deployer/README.md]: mount point is not needed

File: deployer/README.md [hunk 10]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+                - name: varlog
+                  mountPath: /var/log
```
Comment (t0ffel) [inline deployer/README.md]: mount point is not needed

File: deployer/README.md [hunk 11]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+                - name: varlibdockercontainers
+                  mountPath: /var/lib/docker/containers
```
Comment (t0ffel) [inline deployer/README.md]: mount point is not needed

File: deployer/README.md [hunk 12]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+                  value: <logging-es | logging-es-ops>
+                - name: "OPS_PORT"
+                  value: 9200
+                - name: "OPS_CLIENT_CERT"
+                  value: /etc/fluent/keys/cert
+                - name: "OPS_CLIENT_KEY"
+                  value: /etc/fluent/keys/key
+                - name: "OPS_CA"
+                  value: /etc/fluent/keys/ca
+              volumes:
```
Comment (t0ffel) [inline deployer/README.md]: journal and /var/log volumes are not needed

File: deployer/README.md [hunk 13]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+        type: Recreate
+      template:
+        metadata:
+          name: enrichment-fluentd-elasticsearch
+          labels:
+            provider: openshift
+            component: "fluentd"
+        spec:
+          terminationGracePeriodSeconds: 300
+          serviceAccountName: aggregated-logging-fluentd
```
Comment (t0ffel) [inline deployer/README.md]: Ideally it would probably be good to have a separate SA for collectors that are not `cluster-reader`. But I guess that's too much for the purpose..

File: deployer/README.md [hunk 14]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+                volumeMounts:
+                - name: runlogjournal
+                  mountPath: /run/log/journal
+                - name: varlog
+                  mountPath: /var/log
+                - name: varlibdockercontainers
+                  mountPath: /var/lib/docker/containers
+                  readOnly: true
+                - name: config
+                  mountPath: /etc/fluent/configs.d/user
```
Comment (t0ffel) [inline deployer/README.md]: if fluent.conf gets mounted here we'll have:

/etc/fluent/configs.d/user/fluent.conf (which is locally enrichment-fluent.conf),

And this file has a particular include for example:
>    @include configs.d/openshift/system.conf

`/etc/fluent/configs.d/user/configs.d/openshift/system.conf` - shouldn't really exist if i'm not missing anything

File: deployer/README.md [hunk 15]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+                  readOnly: true
+                - name: certs
+                  mountPath: /etc/fluent/keys
+                  readOnly: true
+                - name: dockerhostname
```
Comment (t0ffel) [inline deployer/README.md]: I don't know if all of this hostPath stuff is needed. My guess is that they are not really needed

File: deployer/README.md [hunk 16]
```diff
@@ -1175,3 +1175,260 @@ the first created. You can check if the route in question is defined in multiple
...
+
+    <label @INGRESS>
+      <match **>
+        @type secure_forward
+
+        self_hostname ${HOSTNAME}
+        shared_key <our generated string>
+
+        secure yes
+        ca_cert_path        /etc/fluentd/keys/ca
```
Comment (t0ffel) [inline deployer/README.md]: so the same cert is used to talk between fluentd -> fluentd and fluentd -> ES?

File: deployer/README.md (hunk 1 above)
Comment (ewolinetz) [inline deployer/README.md]: Agreed, that would be a nice thing to be able to configure with our logging role, if this is an approach we'll keep around

File: deployer/README.md (hunk 5 above)
Comment (ewolinetz) [inline deployer/README.md]: Interesting, how would that work though if our only source [for the enrichment pods] is the single `secure-forward` input?

File: deployer/README.md (hunk 6 above)
Comment (ewolinetz) [inline deployer/README.md]: My thought with saying '100' was that even if they wouldn't hit the default limit with maxRequestsInFlight, they would want to re-architect their logging solution to decrease the load it's putting on the API server in general. Would it just be appropriate to call that information out specifically? (2 watches per Fluentd and confirm the default value)

File: deployer/README.md (hunk 7 above)
Comment (ewolinetz) [inline deployer/README.md]: My thought with not providing a template specifically is that we may not make this a permanent solution, so it would be easier to clean up afterwards? I can be convinced otherwise though

File: deployer/README.md (hunk 3 above)
Comment (ewolinetz) [inline deployer/README.md]: I'll just need to make sure I phrase it in a way that they don't use the literal ${SHARED_KEY} like we have for ${HOSTNAME}...

File: deployer/README.md (hunk 8 above)
Comment (ewolinetz) [inline deployer/README.md]: We also want this if we want to do a host port binding, this way we can restrict the number of enrichment fluentd pods to one per node

File: deployer/README.md (hunk 5 above)
Comment (richm) [inline deployer/README.md]: I don't think it would work.  We should not recommend using in_multiprocess unless we can support it, either we provide it or allow the customer to install it, and actually try it out in house and verify exactly what the performance improvement is.  In the testing by @lukas-vlcek he could not figure out how to use it to gain an appreciable performance gain in the openshift aggregated logging case.

File: deployer/README.md (hunk 8 above)
Comment (t0ffel) [inline deployer/README.md]: > if we don't specify that the container runs in the privileged SCC, it won't actually schedule the pod
I don't think so. I believe it'll still schedule the pod - seem to work for me.

> We also want this if we want to do a host port binding, this way we can restrict the number of enrichment fluentd pods to one per node
But why is this needed?

File: deployer/README.md (hunk 7 above)
Comment (t0ffel) [inline deployer/README.md]: I personally find it easier to deal with templates.
For cleanup - if there is a label: "fluentd-role=normalizer" on template, dc and svc then the clean up is much easier.

File: deployer/README.md [hunk 17]
```diff
@@ -1175,3 +1175,249 @@ the first created. You can check if the route in question is defined in multiple
...
+          - name: certs
+            secret:
+              secretName: logging-fluentd
+          - name: dockerhostname
+            hostPath:
+              path: /etc/hostname
+          - name: localtime
+            hostPath:
+              path: /etc/localtime
+          - name: dockercfg
```
Comment (t0ffel) [inline deployer/README.md]: I think this mountpoint is used to understand whether journal log driver is used. It shouldn't be needed here imho.

File: deployer/README.md (hunk 5 above)
Comment (timothysc) [inline deployer/README.md]: 1/10 doesn't seem like that much of a win.  so on a 400 node cluster I need to deploy 40? 

Ideally I would scale with ES shards and schedule the enricher as part of the pod to minimize hops OTW.

Comment (ewolinetz) [inline deployer/README.md]: @timothysc i'm not sure of the ratio we would recommend/use, it may be possible to keep the enricher scale the same as the ES scale. I tried to phrase it with 'likely a good start' trying to defer on that fact. After performance testing has been done with this configuration, a better number can be given

File: deployer/README.md (hunk 8 above)
Comment (ewolinetz) [inline deployer/README.md]: > > if we don't specify that the container runs in the privileged SCC, it won't actually schedule the pod

> I don't think so. I believe it'll still schedule the pod - seem to work for me.

Hmm ok, I'll have to retest that then. I thought that you are prevented from deploying a non-privileged pod with a privileged SA.


> > We  also want this if we want to do a host port binding, this way we can restrict the number of enrichment fluentd pods to one per node

> But why is this needed?

To restrict one enrichment pod to an OpenShift node, otherwise we could end up in a scenario where we have all of our enrichment pods on a single node.

File: deployer/README.md (hunk 16 above)
Comment (ewolinetz) [inline deployer/README.md]: I was unable to get it to work reusing the same certs. Added steps to generate a cert/key pair just for fluentd -> fluentd comunication

**Comment by ewolinetz** [comment]
Made changes based on testing configuration and comments so far

File: deployer/README.md (hunk 7 above)
Comment (ewolinetz) [inline deployer/README.md]: this should still be possible since there are labels on the objects we're creating here `component=enrichment-fluentd`. This works on the svc and dc.

**Comment by ewolinetz** [comment]
Bump -- is there any further discussion that needs to be made regarding this?

**Comment by richm** [comment]
lgtm

File: deployer/README.md [hunk 18]
```diff
@@ -1175,3 +1175,262 @@ the first created. You can check if the route in question is defined in multiple
...
+
+Next, create the key/cert pair that will be used for each Fluentd to communicate
+with one another:
+
+    $ openssl req -new -newkey rsa:2048 -days 712 -x509 -keyout server.key -out server.crt -subj "/CN=logging-enrichment-fluentd/OU=OpenShift/O=Logging" -passout pass:kspass
+
+Once the cert and key are created, the logging-fluentd secret needs to add these
+generated files:
+
+    $ oc create secret generic logging-fluentd --dry-run=true --from-file server.crt --from-file server.key -o yaml | oc apply -f -
```
Comment (sosiouxme) [inline deployer/README.md]: I would put the `dry-run` flag next to the `-o yaml` flag ... first reaction was "oh, he left the --dry-run flag in by accident"

File: deployer/README.md [hunk 19]
```diff
@@ -1175,3 +1175,262 @@ the first created. You can check if the route in question is defined in multiple
...
+
+After that, create the configmap that will be used for the Enriching Fluentd.
+
+First, generate a random key that will be used for communication between the two
+types of Fluentd pods, this will be the value for the `shared_key` configuration
```
Comment (sosiouxme) [inline deployer/README.md]: `s/pods,/pods;/`

File: deployer/README.md (hunk 5 above)
Comment (sosiouxme) [inline deployer/README.md]: How would a user detect if they didn't have enough enrichers for the traffic? Where might they start to see errors, or would log entry just get behind?

File: deployer/README.md [hunk 20]
```diff
@@ -1175,3 +1175,262 @@ the first created. You can check if the route in question is defined in multiple
...
+
+Then, create the configmap that will be used for the Enriching Fluentd, using the
+file created in the previous step as the source for the configmap:
+
+    $ oc create configmap logging-fluentd-enrichment \
+        --from-file=fluent.conf=enrichment-fluent.conf
+
+The next step will be to create the Deployment configuration that will be used to
+scale up the enrichment fluentd pods. Save the following file to your local file
+system as `enrichment-fluentd-dc.yaml`. Be sure to update the value of "K8S_HOST_URL"
```
Comment (sosiouxme) [inline deployer/README.md]: I wouldn't call out `K8S_HOST_URL`, it shouldn't need editing.

File: deployer/README.md (hunk 5 above)
Comment (ewolinetz) [inline deployer/README.md]: @lukas-vlcek @t0ffel is there any messages that we would see in Fluentd to indicate that it isn't able to keep up with the influx of events to the enrichment pods? Either by looking at the logs of the enrichment pods or of the ingress pods?

Would this be a good use of monitoring agent? http://docs.fluentd.org/articles/monitoring

**Comment by t0ffel** [comment]
My only essential concern is that the pod is privileged, while the pod doesn't do anything that should require it to be privileged.

**Comment by ewolinetz** [comment]
It would need to be privileged if its going to do a host port binding though, or should we not care if we have more than one enricher pod on the same node? I would think that could be a performance impact we'd want to avoid

**Comment by richm** [comment]
We may want to have more than one enricher pod running on a node.  fluentd is not threaded and does not scale well for multiple CPUs.  If you have a node with multiple CPUs, you might actually achieve high CPU utilization by running multiple fluentd on the node.  And this will work with the enricher case (as oppposed to the collector case where it is difficult to run multiple fluentd and/or in_multiprocess because you can't easily divide reading data form the different sources among multiple processes).

**Comment by ewolinetz** [comment]
@richm works for me. @t0ffel I'll verify that reusing the privileged SA with an unprivileged pod works and make the change

**Comment by ewolinetz** [comment]
@richm @t0ffel removed privileged pod sections from enrichment DC, verified we can scale up to more than one replica on a host

File: deployer/README.md (hunk 5 above)
Comment (lukas-vlcek) [inline deployer/README.md]: @ewolinetz I believe this is probably plugin specific. There might be some common configuration custom followed by all supported plugins (supported by core Fluentd develoeprs), but for community plugins I am afraid this is specific.

For example I realised that for Elasticsearch output plugin all responses from ES that are not `200` are ignored (i.e. dropped). This means that all bulk indexing requests (from Fluentd to ES) are dropped if they are rejected by ES. So there does not seem to be any Fluentd core functionality that plugins can re-use in such cases? See https://github.com/uken/fluent-plugin-elasticsearch/issues/105

As you mentioned it is worth investigating if Fluentd monitoring REST API can give us some information about back pressure or lack of processing resources. When I was investigating the above mentioned issue with ES bulk indexing I checked the this REST API (there is example of REST API output in the end of this comment https://github.com/ViaQ/integration-tests/issues/6#issuecomment-227431454) and is can expose info about internal queues. Again, not sure if other plugins expose the same information. But since the secure forward should be core supported plugin we could hope...

HTH

Comment (ewolinetz) [inline deployer/README.md]: So far with just one enricher and one collector, I added the monitoring endpoint and using `oc exec <pod> -- -- curl http://localhost:24220/api/plugins.json | python -m json.tool | grep buffer` I could see the buffer_total_queued_size increase and then go back to 0.

That may be a good indicator whether or not the enrichment pod is able to keep up, if we see our collector pods not be able to fully flush their buffers and the enrichment pods getting backed up in their queues.

For some reason on my system the non-operations ES plugin on the enricher pod isn't able to fully flush, same with the collector (this may be due to the enricher pod not fully flushing).

I believe the correct pattern, if this were to be graphed, would be a "saw" shape. If we see linear growth with that it, that may be an indication that we aren't able to keep up. I'll try to resolve this buffer not clearing (I don't see any errors in the Fluentd logs), and see if my theory holds true.

**Comment by ewolinetz** [comment]
Added information on monitoring the Fluentd buffer usage

**Comment by jcantrill** [comment]
@ewolinetz Did we ever produce this kbase?  I believe the info is probably still relevant.

**Comment by portante** [comment]
@jcantrill, @richm, isn't this replaced by mux?  Would it be better to just point them at the mux playbook?

**Comment by richm** [comment]
@portante yes, this is essentially what turned into mux

**Comment by jcantrill** [comment]
@richm so we should just close this PR or is it still a viable kbase we wish to publish?

**Comment by richm** [comment]
I think we should publish this for 3.5/1.5 and earlier releases

**Comment by jcantrill** [comment]
@ewolinetz possible to fix this up with the current stack and add to our docs here?

**Comment by jcantrill** [comment]
Closing now.  Resolvable in 4.x via pipelines when implemented.
